atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketCloseTimestampCalculator.java

```diff
@@ -0,0 +1,148 @@
+        if (finalLogicalTimestamp <= startTimestamp) {
+            long freshTimestamp = freshTimestampSupplier.get();
+            if (freshTimestamp - startTimestamp < MIN_BUCKET_SIZE_FOR_NON_PUNCHER_CLOSE) {
+                log.info(
+                        NON_PUNCHER_CLOSE_LOG
+                                + " We read a fresh timestamp {}, but this is not sufficiently far from the start"
+                                + " timestamp to close the bucket.",
+                        SafeArg.of("finalTimestampFromPunchTable", finalLogicalTimestamp),
+                        SafeArg.of("startTimestamp", startTimestamp),
+                        SafeArg.of("timeGap", TIME_GAP_BETWEEN_BUCKET_START_AND_END),
+                        SafeArg.of("openWallClockTimeMillis", openWallClockTimeMillis),
+                        SafeArg.of("freshTimestamp", freshTimestamp));
+                return CloseableState.notCloseable();
+            }
+
+            long cappedTimestamp = Math.min(startTimestamp + MAX_BUCKET_SIZE_FOR_NON_PUNCHER_CLOSE, freshTimestamp);
+            if (cappedTimestamp != freshTimestamp) {
+                log.info(
+                        NON_PUNCHER_CLOSE_LOG
+                                + " We read a fresh timestamp {}, but this is too far from the start timestamp."
+                                + " Proposing a capped timestamp {} instead.",
+                        SafeArg.of("finalTimestampFromPunchTable", finalLogicalTimestamp),
+                        SafeArg.of("startTimestamp", startTimestamp),
+                        SafeArg.of("timeGap", TIME_GAP_BETWEEN_BUCKET_START_AND_END),
+                        SafeArg.of("openWallClockTimeMillis", openWallClockTimeMillis),
+                        SafeArg.of("freshTimestamp", freshTimestamp),
+                        SafeArg.of("cappedTimestamp", cappedTimestamp));
+            } else {
+                log.info(
+                        NON_PUNCHER_CLOSE_LOG
+                                + " We read a fresh timestamp {}, and this will be the proposed closing timestamp",
+                        SafeArg.of("finalTimestampFromPunchTable", finalLogicalTimestamp),
+                        SafeArg.of("startTimestamp", startTimestamp),
+                        SafeArg.of("timeGap", TIME_GAP_BETWEEN_BUCKET_START_AND_END),
+                        SafeArg.of("openWallClockTimeMillis", openWallClockTimeMillis),
+                        SafeArg.of("freshTimestamp", freshTimestamp));
+            }
+            return CloseableState.closeable(cappedTimestamp);
+        } else {
+            return CloseableState.closeable(finalLogicalTimestamp);
+        }
+    }
+
+    @Value.Immutable
+    public interface CloseableState {
+        @Value.Parameter
+        boolean canBeClosed();
+
+        // Only relevant if canBeClosed is true.
+        // TODO: Should this be a proper ADT? Seems overkill.
```

Comment:
Q for reviewer, I think it's overkill for this unlike the bucket assigner, but open to thoughts as always.

Reply:
ADT is certainly overkill. I'm not sure if this is preferable to wrapping `OptionalLong` to be honest, since `true` always has a timestamp and `false` never has one?